app/routes.py

```diff
@@ -11,131 +11,143 @@
 from app import app
 from app.helpers.dynamodb import get_db
 from app.helpers.s3 import get_storage
+from app.helpers.utils import get_kml_file_link
+from app.helpers.utils import validate_content_length
 from app.helpers.utils import validate_content_type
-from app.helpers.utils import validate_kml_string
-from app.settings import KML_STORAGE_URL
+from app.helpers.utils import validate_kml_file
+from app.helpers.utils import validate_permissions
+from app.settings import KML_MAX_SIZE
+from app.settings import ROUTE_ADMIN_PREFIX
+from app.settings import ROUTE_BASE_PREFIX
+from app.settings import ROUTE_FILES_PREFIX
 from app.version import APP_VERSION
 
 logger = logging.getLogger(__name__)
 
 
-@app.route('/checker', methods=['GET'])
-def check():
+@app.route(f'/{ROUTE_BASE_PREFIX}/checker', methods=['GET'])
+def checker():
     return make_response(jsonify({'success': True, 'message': 'OK', 'version': APP_VERSION}))
 
 
-@app.route('/kml', methods=['POST'])
-@validate_content_type("application/vnd.google-earth.kml+xml")
-def post_kml():
-    # IE9 sends data urlencoded
-    quoted_str = request.get_data().decode('utf-8')
-    data = unquote_plus(quoted_str)
-    kml_string = validate_kml_string(data)
-    kml_admin_id = base64.urlsafe_b64encode(uuid.uuid4().bytes).decode('utf8').replace('=', '')
-    kml_id = base64.urlsafe_b64encode(uuid.uuid4().bytes).decode('utf8').replace('=', '')
-    timestamp = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc).isoformat()
+# NOTE the /kml/files/<kml_id> route is directly served by S3
+
+
+@app.route(f'/{ROUTE_ADMIN_PREFIX}', methods=['POST'])
+@validate_content_type("multipart/form-data")
+@validate_content_length(KML_MAX_SIZE)
+def create_kml():
+    # Get the kml file data
+    kml_string, empty = validate_kml_file()
+
+    kml_admin_id = urlsafe_b64encode(uuid4().bytes).decode('utf8').replace('=', '')
+    kml_id = urlsafe_b64encode(uuid4().bytes).decode('utf8').replace('=', '')
+    file_key = f'{ROUTE_FILES_PREFIX}/{kml_id}'
+    timestamp = datetime.utcnow().replace(tzinfo=timezone.utc).isoformat()
+
     storage = get_storage()
-    storage.upload_object_to_bucket(kml_id, kml_string)
-    db = get_db()  # pylint: disable=invalid-name
-    db.save_item(kml_admin_id, kml_id, timestamp)
-    item = db.get_item(kml_admin_id)
+    storage.upload_object_to_bucket(file_key, kml_string)
+
+    db = get_db()
+    db.save_item(kml_id, kml_admin_id, file_key, timestamp, empty)
```

Comment:
Might make sense to only store the kml in case it is not empty?

Reply:
@hansmannj the `empty` feature has been taken from mf-chsdi3 and an empty kml is not an empty file, but a file with `<kml></kml>`, so rather an empty kml xml element. Not sure if we wil use this with the new viewer but might be useful. The goal of this empty kml element was for permalink that would otherwise point to a non existing file.